src/App/mermaidData/pullApiData.js

```diff
@@ -44,80 +62,111 @@ export const pullApiData = async ({
+              shouldHandleUserDeniedSyncPull = true
+            }
+
+            const isDataTypeProjectAssociated = getIsDataTypeProjectAssociated(apiDataType)
+
+            const bulkDeleteIdsWithNoDuplicates = Array.from(new Set([...deleteIds, ...removeIds]))
+
+            dexiePerUserDataInstance[apiDataType].bulkPut(updatesWithPushToApiTagReset)
+            dexiePerUserDataInstance[apiDataType].bulkDelete(bulkDeleteIdsWithNoDuplicates)
+
+            if (is401Or403 && isDataTypeProjectAssociated && projectId) {
+              // we still delete project related data in addition to anything in the removes array,
+              // because the removes array response doesnt include all the things we want to delete currently.
+
+              shouldDeleteProjectFromBeingConsideredOfflineReady = true
+              const deleteProjectRelatedRecords = dexiePerUserDataInstance[apiDataType]
+                .where({ project: projectId })
+                .delete()
+              const resetProjectRelatedLastRevisionNumbers =
+                resetLastRevisionNumberForProjectDataType({
+                  dataType: apiDataType,
+                  projectId,
+                  dexiePerUserDataInstance,
+                })
+
+              await Promise.all([
+                deleteProjectRelatedRecords,
+                resetProjectRelatedLastRevisionNumbers,
+              ])
+            }
           }
+        })
+
+        if (shouldHandleUserDeniedSyncPull && projectId && projectName) {
+          /* user will only be denied pulling when they are trying to
+           pull from a entity where there will also be a projectId
+           as part of the pull request, which we need to determine projectName
+           we do that at the beginning of the pullApiData funciton instead of down here,
+           because by the time we get to this block, the project info with its project name are
+           likely to have already been deleted with the content of the removes array */
+
+          handleUserDeniedSyncPull(projectName)
         }
-      })
-    },
-  )
+        if (shouldDeleteProjectFromBeingConsideredOfflineReady && projectId) {
+          /* we wait to delete the project from being considered offline ready
+           until after we call handleUserDeniedSyncPull which involves a redirect away from the
+           a project specific page (crudly using a setTimeout). This is because if the app
+           is on a project-specific page, it automatically sets the project as offline-ready
+           (as a business feature), we need to ensure that the redirect
+           to happen first so that us deleting the project from being offline - ready persists.
+           We want to avoid the project just being reset as offline-ready
+           immediately after we remove it from the offline-ready list of projects.
+           This pollutes this pull logic and module with some dependence
+           on the UI code and routing logic. */
+
```

Comment:
Ugh, I wish history.push returned a promise once the route was done changing. Using setTimeout instead. This will slow down pull syncs when there are subnode/nested 403s which I think is ok since those are an edge-ish case or at least not happy path.